### PR TITLE
Allow PLY to generate parsing/lexing tables.

### PR DESCRIPTION
**Pull request by FLHerne**
...and modify setup.py to create them at build time.

On my laptop, pre-generated tables reduce the time spent in
 `NMLParser.__init__()` from 2.2s to 0.2s, halving the total runtime
 of nmlc for small input files.

PLY will attempt to regenerate these automatically at runtime if the
 grammar has changed. In the case where they're system-installed and not
 writeable, the grammar is also system-installed and not writeable. :P

### CODE REVIEW

**Comment by glx22** [comment]
A small change is required for pyinstaller. I made the changes locally and can push them to your branch if it's ok.

**Comment by FLHerne** [comment]
@glx22 Yes, please go ahead. :-)

**Comment by andythenorth** [comment]
I tested this with Iron Horse and  --verbosity=4 flag.  Python is 3.5.

Results:
- Init parser: consistently 0.0s with this PR, 0.4-0.5s in master
- Parsing: less consistent results, likely due to other system load.  With this PR 16.4-17.3s.  Without this PR 17.0-20.0s.

**Comment by glx22** [comment]
Hmm `pip install -e .` or `python setup.py develop` fails with the setup.py changes if ply is not yet installed. Without the changes the dependencies are installed and everything is fine.

File: setup.py [hunk 1]
```diff
@@ -3,9 +3,18 @@
 import os
 
 from setuptools import setup, Extension, find_packages
-from nml import version_info
+from setuptools.command.build_py import build_py
+from nml import parser, version_info
```
Comment (glx22) [inline setup.py]:
```suggestion
from nml import version_info
```

File: setup.py [hunk 2]
```diff
@@ -3,9 +3,18 @@
...
 NML_VERSION = version_info.get_and_write_version()
 
+
+class NMLBuildPy(build_py):
+    def run(self):
+        # Create a parser so that nml/generated/{parse,lex}tab.py are generated.
```
Comment (glx22) [inline setup.py]:
```suggestion
        from nml import parser
        # Create a parser so that nml/generated/{parse,lex}tab.py are generated.
```

Comment (FLHerne) [reply setup.py]: Thanks. This fix was what I thought of when I saw your last comment, but I hadn't got to it yet. :P

**Pull request merged**